Merge qa_loader and create-dev-data; add seen data.

rmol opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes #5514.

There was a lot of overlap between `qa_loader.py` and `create-dev-data.py`, so we're consolidating them while adding seen data to the populated database.

## Testing

- Run `make dev` and verify that the database is populated correctly.
- Run `make build-debs && make staging`, log in to the app server, and run:
  - `sudo -u www-data bash`
  - `cd /var/www/securedrop`
  - `CRYPTOGRAPHY_ALLOW_OPENSSL_102=yes ./loaddata.py --source-count ALL`. 
  The default three journalists and a number of sources should be created without any errors.
  - Run `loaddata.py` a few more times, varying the number of journalists, sources, and files, messages and replies per source, to be created. Confirm the results are as expected.

## Deployment

This is dev/test only.

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

Choose one of the following:

- [x] I have opened https://github.com/freedomofpress/securedrop-docs/pull/108 in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes
- [ ] I would appreciate help with the documentation
- [ ] These changes do not require documentation

rmol committed:
Merge qa_loader and create-dev-data; add seen data.

There was a lot of overlap between qa_loader.py and
create-dev-data.py, so we're consolidating them.

zenmonkeykstop commented:
Couple of issues as per below:

## dev 
- [x] Run make dev and verify that the database is populated correctly. ***Only 2 journalists left of 3 created - as expected as one is deleted after the source replies are created**

## staging
- [x] The default three journalists and a number of sources should be created without any errors. **same 2 journalists created.**
- [ ] Run loaddata.py a few more times, varying the number of journalists, sources, and files, messages and replies per source, to be created. Confirm the results are as expected. **subsequent runs just add extra records instead of resetting database first - this could be addressed in documentation by asking dev to run `./manage.py reset` between runs.**

sssoleileraaa reviewed (dismissed):
`make dev` shows all three sources as unseen now. i think it would be useful to go back to showing the default sources with more variation, e.g. the way it is on `develop`:
1. one source with all conversation items seen
2. one source with all conversation items unseen
3. one source with some conversation items seen (so the source will show up as unseen)

also, it looks like you show some sources without replies. 

this is how it looks like on develop when you run `NUM_SOURCES=10 make dev`: 
![Screenshot from 2020-11-24 18-21-33](https://user-images.githubusercontent.com/4522213/100174799-020d8a00-2e82-11eb-9659-c3a39e634ede.png)

now it looks like:
![Screenshot from 2020-11-24 18-23-30](https://user-images.githubusercontent.com/4522213/100174896-3123fb80-2e82-11eb-9c89-e87c303c808a.png)

did you intend to change this?

zenmonkeykstop commented:
@creviera I got 2 unseen sources and one seen on my `make dev` run, I think that's just down to the roll of the dice. On the replies front, it looks like once the fraction of sources that are supposed to get replies has been reached, all others afterwards get zero.

zenmonkeykstop commented on securedrop/loaddata.py:
```diff
@@ -0,0 +1,419 @@
+            source, codename = add_source()
+
+            for _ in range(args.messages_per_source):
+                journalist_who_saw = random.choice(
```
@creviera - this is why seen/unseen varies

sssoleileraaa replied:
yup, random.choice is not going to be predictable

sssoleileraaa replied:
i think this pr should try to keep the current behavior with seen/unseen made in the PR: https://github.com/freedomofpress/securedrop/pull/5580

i use this script everyday so it's nice to have predictability with creating seen data, although perhaps we should have more variation with file attachments and last_updated, but this can be done outside of this pr.

zenmonkeykstop commented on securedrop/loaddata.py:
```diff
@@ -0,0 +1,419 @@
+            if i <= starred_sources_count:
+                star_source(source)
+
+            reply_to_source = i <= replied_sources_count
```
@creviera - this is why no replies after a certain point (can get 100% by setting reply fraction arg to 1)

sssoleileraaa replied:
good find

zenmonkeykstop reviewed (dismissed):
Script populates test data as expected, with the exception of the Death Of Superman. One worthwhile change would be to reset the database before loading data, though this could also be addressed by documenting the existing manage.py action for this.

zenmonkeykstop committed:
Added random number seed to increase dataset consistency

sssoleileraaa reviewed (changes requested):
:heavy_check_mark: cassette generation for seen/unseen in the client is working again
:heavy_check_mark: a couple client functional tests (that are more brittle) are failing during cassette generation 
  - there is now 1 file attachment instead of 2 when running `make dev` which affects some functional tests, e.g. `test_file_download` in the client
  - this is fine (not great but fine) since CI will not fail on the client repo. Client functional tests can be updated any time after this is merged to work with the new default config

:x: some SDK functional tests (that are more brittle) are failing during cassette generation
  - now, not every source has a reply be default when running `make dev` which affects some tests, e.g. `test_get_all_replies` and `test_get_replies_from_source` in the sdk
  - there will be CI failures since we have an SDK CI job that tests cassette generation. I think if we just add replies to each source by default again (the way it is on `main`) when running `make dev`, the SDK functional tests will work again, but it would be best to make sure by following these steps to test cassette generation: https://github.com/freedomofpress/securedrop-sdk#generating-cassettes-for-api-calls-over-http. Another idea: If you don't want to add replies to every source by default, then you'll have to make a PR in the SDK repo with the functional test changes needed in order to work with this new default config. This will need to go out fairly quickly after this PR is merged and will take some time, so it should be done in parallel.

rmol commented:
The deleted journalist is expected; sorry for not noting that in the test plan.

There's a couple of axes of tension here: between dev and staging (the current `create-dev-data.py` and `qa_loader.py` scripts), and between randomness and predictability in the test dataset.

We shouldn't fix the seed by default when running `loaddata.py` manually, because when you're iteratively populating a database, perhaps for load testing at various sizes, it's not obvious that the duplicate records are caused by the fixed seed (#5236).

That same scenario is why `loaddata.py` shouldn't reset the database.

I think we should try to have less predictability in the dev and test database by default. Tests that don't rely on cassettes should either establish the state they need (which could be done with fixtures -- #3836) or work with what they find, without brittle expectations. It's more work, but it's more robust, and it'll be important when we try to tackle #2535. I've also found variable data useful in interactive use; when changing the UI it can turn up things you didn't account for.

To support fixture generation, I could modify `dev-shell`, `dev-deps` and the `Makefile` to add a controlled configuration, always using the same seed and counts, and have the SDK CI job use that.

Thoughts?

sssoleileraaa commented:
That makes sense, John. What do you think about doing things in the following order (this PR should tackle the first step):

1. Merge the two scripts into one without breaking CI and ideally without breaking any cassette generation test in the client or in the SDK. (Currently, cassette generation is broken with the changes in this PR.)
2. Change the functional tests that would break based on the changes you want to make.
3. Make changes such as adding ways to configure how we initialize the dev server, e.g. sometimes we want each initialization to look the same during UI and UX development so we can meticulously compare, sometimes we want data that looks more like a real server for live demos, sometimes we want a lot of randomness and variation to catch issues we might not see otherwise. (This is a good topic for our engineering meeting.)
4. More changes to how we do functional tests and cassette generation

rmol committed:
More loaddata.py configuration

A few tests in securedrop-sdk and securedrop-client make assumptions
about the content of the test database, so until they can be fixed,
make sure the default result of loaddata.py doesn't break them.

Change loaddata.py to not fix the random seed by default, as this
breaks additive usage, as in a staging environment for load testing,
in a non-obvious way.

Add the --seen-message-fraction and --seen-file-fraction to
loaddata.py, giving more control over the number of submissions marked
seen.

Clarify loaddata.py output.

Add the ability to pass loaddata.py options through dev-shell via the
LOADDATA_ARGS environment variable. Remove the NUM_SOURCES and
NUM_JOURNALISTS options from dev-shell.

rmol commented:
I've changed the `loaddata.py` defaults to match the expectations of the client and SDK tests. I tested cassette generation in `securedrop-sdk` and had no trouble.

I've changed `dev-shell` to support passing the `LOADDATA_ARGS` environment variable, which `dev-deps` will use if set. This will let us lock down parameters in things like the SDK cassette generation CI test, so that down the road we can change the default test dataset as we like without affecting those.

rmol committed:
Restore NUM_JOURNALISTS and NUM_SOURCES

rmol committed:
After rebase, reremove create-dev-data.py and qa_loader.py

rmol commented:
Rebased and ready for review again.

rmol committed:
loaddata.py: Add missed type annotation in record_source_interaction

sssoleileraaa reviewed (approved):
lgtm
